plugin/src/main/java/org/opensearch/ml/action/update/UpdateModelTransportAction.java

```diff
@@ -0,0 +1,564 @@
+        if (newModelGroupId != null) {
+            modelAccessControlHelper
+                .validateModelGroupAccess(user, newModelGroupId, client, ActionListener.wrap(hasNewModelGroupPermission -> {
+                    if (hasNewModelGroupPermission) {
+                        mlModelGroupManager.getModelGroupResponse(newModelGroupId, ActionListener.wrap(newModelGroupResponse -> {
+                            updateRequestConstructor(
+                                modelId,
+                                newModelGroupId,
+                                updateRequest,
+                                updateModelInput,
+                                newModelGroupResponse,
+                                wrappedListener,
+                                isUpdateModelCache,
+                                isPredictorUpdate
+                            );
+                        },
+                            exception -> wrappedListener
+                                .onFailure(
+                                    new OpenSearchStatusException(
+                                        "Failed to find the model group with the provided model group id in the update model input, MODEL_GROUP_ID: "
+                                            + newModelGroupId,
+                                        RestStatus.NOT_FOUND
+                                    )
+                                )
+                        ));
+                    } else {
+                        wrappedListener
+                            .onFailure(
+                                new OpenSearchStatusException(
+                                    "User Doesn't have privilege to re-link this model to the target model group due to no access to the target model group with model group ID "
+                                        + newModelGroupId,
+                                    RestStatus.FORBIDDEN
+                                )
+                            );
+                    }
+                }, exception -> {
+                    log.error("Permission denied: Unable to update the model with ID {}. Details: {}", modelId, exception);
+                    wrappedListener.onFailure(exception);
+                }));
+        } else {
+            updateRequestConstructor(modelId, updateRequest, updateModelInput, wrappedListener, isUpdateModelCache, isPredictorUpdate);
+        }
+    }
+
+    private void updateRequestConstructor(
+        String modelId,
+        UpdateRequest updateRequest,
+        MLUpdateModelInput updateModelInput,
+        ActionListener<UpdateResponse> wrappedListener,
+        boolean isUpdateModelCache,
+        boolean isPredictorUpdate
+    ) {
+        try {
+            updateModelInput.setLastUpdateTime(Instant.now());
+            updateRequest.doc(updateModelInput.toXContent(XContentFactory.jsonBuilder(), ToXContent.EMPTY_PARAMS));
+            updateRequest.docAsUpsert(true);
```

Comment:
Should we set refresh policy as immediate?

Reply:
Added to all